File: DriverDrowsinessDetectorApp/app/src/main/java/com/example/driverdrowsinessdetectorapp/data/remote/api/EventosApi.kt
package com.example.driverdrowsinessdetectorapp.data.remote.api

import com.example.driverdrowsinessdetectorapp.data.remote.dto.request.EventoRequest
import com.example.driverdrowsinessdetectorapp.data.remote.dto.request.EventosBatchRequest
import com.example.driverdrowsinessdetectorapp.data.remote.dto.response.EstadisticasChoferResponse
import com.example.driverdrowsinessdetectorapp.data.remote.dto.response.EventoResponse
import com.example.driverdrowsinessdetectorapp.data.remote.dto.response.EventoResumenResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

/**
 * Interface Retrofit para endpoints de eventos de somnolencia.
 * 
 * Base URL: /api/v1/eventos
 * 
 * Todos los endpoints requieren autenticación JWT Bearer Token.
 * 
 * @author Sistema de Detección de Somnolencia
 * @version 1.0
 */
interface EventosApi {
    
    // ENDPOINTS PARA CREAR EVENTOS (CHOFER)
    
    /**
     * Crear un evento de somnolencia individual.
     * 
     * Usado cuando hay conexión a internet y se puede enviar inmediatamente.
     * 
     * @param evento Datos del evento a crear
     * @return Evento creado con ID asignado por el servidor
     */
    @POST("/api/v1/eventos")
    suspend fun crearEvento(
        @Body evento: EventoRequest
    ): Response<EventoResponse>
    
    /**
     * Crear múltiples eventos en lote (sincronización offline).
     * 
     * Usado cuando el chofer recupera conexión después de estar offline
     * y necesita sincronizar eventos acumulados en Room.
     * 
     * Límites:
     * - Mínimo: 1 evento
     * - Máximo: 100 eventos por request
     * 
     * @param batch Lista de eventos a sincronizar
     * @return Lista de eventos creados con IDs asignados
     */
    @POST("/api/v1/eventos/batch")
    suspend fun crearEventosBatch(
        @Body batch: EventosBatchRequest
    ): Response<List<EventoResponse>>
    
    // ENDPOINTS PARA CONSULTAR EVENTOS
    
    /**
     * Obtener mis eventos (para chofer).
     * 
     * Permite al chofer ver su historial de eventos.
     * 
     * @param limite Número máximo de resultados (1-200, default: 50)
     * @param offset Offset para paginación (default: 0)
     * @param dias Eventos de los últimos X días (1-90, default: 7)
     * @param tipoEvento Filtrar por tipo de evento (opcional)
     * @param nivelSeveridad Filtrar por severidad (opcional)
     * @return Lista de eventos resumidos
     */
    @GET("/api/v1/eventos/mis-eventos")
    suspend fun getMisEventos(
        @Query("limite") limite: Int = 50,
        @Query("offset") offset: Int = 0,
        @Query("dias") dias: Int = 7,
        @Query("tipo_evento") tipoEvento: String? = null,
        @Query("nivel_severidad") nivelSeveridad: String? = null
    ): Response<List<EventoResumenResponse>>
    
    /**
     * Obtener eventos de un viaje específico.
     * 
     * @param idViaje ID del viaje
     * @return Lista de eventos del viaje
     */
    @GET("/api/v1/eventos/viaje/{id_viaje}")
    suspend fun getEventosPorViaje(
        @Path("id_viaje") idViaje: Int
    ): Response<List<EventoResumenResponse>>
    
    // ENDPOINTS DE ESTADÍSTICAS
    
    /**
     * Obtener mis estadísticas (para chofer).
     * 
     * @param dias Estadísticas de los últimos X días (1-90, default: 7)
     * @return Estadísticas del chofer
     */
    @GET("/api/v1/eventos/estadisticas/mis-estadisticas")
    suspend fun getMisEstadisticas(
        @Query("dias") dias: Int = 7
    ): Response<EstadisticasChoferResponse>
}

File: drowsiness-detection-app/src/features/eventos/services/eventosApi.ts
/**
 * Servicio API para Eventos de Somnolencia
 */
import apiClient from '../../../lib/api/client';
import type { 
  EventoConChofer, 
  EventoResumen,
  EstadisticasGenerales, 
  EstadisticasChofer,
  EventosFiltros,
  TipoEvento,
  NivelSeveridad
} from '../types';

const BASE_URL = '/eventos';

export const eventosApi = {
  /**
   * Obtener eventos recientes (para dashboard admin)
   */
  getEventosRecientes: async (params: {
    minutos?: number;
    limite?: number;
    solo_criticos?: boolean;
  } = {}): Promise<EventoConChofer[]> => {
    try {
      const response = await apiClient.get<EventoConChofer[]>(`${BASE_URL}/recientes`, {
        params: {
          minutos: params.minutos || 60,
          limite: params.limite || 100,
          solo_criticos: params.solo_criticos || false
        }
      });
      return response.data;
    } catch (error: any) {
      console.error('Error en getEventosRecientes:', error.response?.data || error.message);
      throw error;
    }
  },

  /**
   * Obtener eventos de un chofer específico
   */
  getEventosPorChofer: async (
    idChofer: number,
    params: {
      limite?: number;
      offset?: number;
      dias?: number;
      tipo_evento?: TipoEvento;
      nivel_severidad?: NivelSeveridad;
    } = {}
  ): Promise<EventoResumen[]> => {
    try {
      const response = await apiClient.get<EventoResumen[]>(`${BASE_URL}/chofer/${idChofer}`, {
        params: {
          limite: params.limite || 50,
          offset: params.offset || 0,
          dias: params.dias || 7,
          ...(params.tipo_evento && { tipo_evento: params.tipo_evento }),
          ...(params.nivel_severidad && { nivel_severidad: params.nivel_severidad })
        }
      });
      return response.data;
    } catch (error: any) {
      console.error('Error en getEventosPorChofer:', error.response?.data || error.message);
      throw error;
    }
  },

  /**
   * Obtener eventos de un viaje
   */
  getEventosPorViaje: async (idViaje: number): Promise<EventoResumen[]> => {
    try {
      const response = await apiClient.get<EventoResumen[]>(`${BASE_URL}/viaje/${idViaje}`);
      return response.data;
    } catch (error: any) {
      console.error('Error en getEventosPorViaje:', error.response?.data || error.message);
      throw error;
    }
  },

  /**
   * Obtener estadísticas generales del sistema
   */
  getEstadisticasGenerales: async (dias: number = 7): Promise<EstadisticasGenerales> => {
    try {
      const response = await apiClient.get<EstadisticasGenerales>(
        `${BASE_URL}/estadisticas/generales`,
        { params: { dias } }
      );
      return response.data;
    } catch (error: any) {
      console.error('Error en getEstadisticasGenerales:', error.response?.data || error.message);
      throw error;
    }
  },

  /**
   * Obtener estadísticas de un chofer específico
   */
  getEstadisticasChofer: async (
    idChofer: number,
    dias: number = 7
  ): Promise<EstadisticasChofer> => {
    try {
      const response = await apiClient.get<EstadisticasChofer>(
        `${BASE_URL}/estadisticas/chofer/${idChofer}`,
        { params: { dias } }
      );
      return response.data;
    } catch (error: any) {
      console.error('Error en getEstadisticasChofer:', error.response?.data || error.message);
      throw error;
    }
  },

  /**
   * Obtener todos los eventos con filtros (para exportación)
   */
  getAllEventos: async (filtros: EventosFiltros = {}): Promise<EventoConChofer[]> => {
    // Usar endpoint de recientes con límite alto
    const minutos = (filtros.dias || 30) * 24 * 60; // Convertir días a minutos
    
    return eventosApi.getEventosRecientes({
      minutos,
      limite: filtros.limite || 1000,
      solo_criticos: filtros.solo_criticos
    });
  }
};

export default eventosApi;